lcls-twincat-motion/Library/POUs/Motion/FB_GantryAutoCoupling.TcPOU

```diff
@@ -0,0 +1,89 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4020.14">
+  <POU Name="FB_GantryAutoCoupling" Id="{532dfcaf-6187-4140-aaad-e20a9bbe6b43}" SpecialFunc="None">
+    <Declaration><![CDATA[FUNCTION_BLOCK FB_GantryAutoCoupling
+VAR_INPUT
+	fGantryTol : LREAL;
+END_VAR
+VAR_OUTPUT
+	bGantryAlreadyCoupled : BOOL;
+END_VAR
+VAR_IN_OUT
+	Master : DUT_MotionStage;
+	MasterEnc : ST_RenishawAbsEnc;
+	Slave : DUT_MotionStage;
+	SlaveEnc : ST_RenishawAbsEnc;
+	bExecuteCouple : BOOL;
+	bExecuteDecouple : BOOL;
+END_VAR
+VAR
+	gantry_diff_limit : FB_GantryDiffVirtualLimitSwitch;
+	couple : MC_GEARIN;
+	decouple : MC_GEAROUT;
+
+
+	bInitComplete : BOOL;
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[// Designate Master and SLave Axes
+Master.bGantryAxis := TRUE;
+Slave.bGantryAxis := TRUE;
+
+Master.fGantryTol := fGantryTol;
+Slave.fGantryTol := Master.fGantryTol;
+
+// Activate Gantry Virtual Limit Switch
+gantry_diff_limit(Penc:=MasterEnc, SEnc:=SlaveEnc, GantDiffTol:=LREAL_TO_LINT(Master.fGantryTol),
+				  PLimFwd=>Master.bLimGntFwd, PLimBwd=>Master.bLimGntBwd,
+				  SLimFwd=>Slave.bLimGntFwd, SLimBwd=>Slave.bLimGntBwd);
+
+// Coupling Status Bit
+bGantryAlreadyCoupled := Master.Axis.NcToPlc.CoupleState=1 AND Slave.Axis.NcToPlc.CoupleState=3;
+
+// Make master and slave axes respect each other's limits
+IF bGantryAlreadyCoupled THEN
+	// Fwd limits
+	Master.bLimGntFwd := Master.bLimGntFwd AND Slave.bLimGntFwd AND Slave.bLimFwd;
+	Slave.bLimGntFwd := Master.bLimGntFwd AND Slave.bLimGntFwd AND Master.bLimFwd;
+	// Bwd limits
+	Master.bLimGntBwd := Master.bLimGntBwd AND Slave.bLimGntBwd AND Slave.bLimBwd;
+	Slave.bLimGntBwd := Master.bLimGntBwd AND Slave.bLimGntBwd AND Master.bLimBwd;
```

Comment:
Is there a reason why we can't use the more complete `AND Slave.bAllFowardEnable`, etc. variables? Are we not guaranteed that `FB_MotionStage` is running for the slave axis?

Reply:
Now that you say it I don't have a reason why we aren't using the `bAllForwardEnable` variables. I believe we are guaranteed that the slave still uses `FB_MotionStage`, and this seems like the same logic but would be much easier to read. I'll try now and confirm

Reply:
I'm actually not so sure about this. The main issue that led to this code snippet is the fact that if a Slave's limit switch activates before a master's, motion will continue for both axes. I tried the following:
```
IF bGantryAlreadyCoupled THEN
	Master.bLimAllFwd := Master.bLimAllFwd AND Slave.bLimAllFwd;
	Master.bLimAllBwd := Master.bLimAllBwd AND Slave.bLimAllBwd;
	
	Slave.bLimAllFwd := Master.bLimAllFwd;
	Slave.bLimAllBwd := Master.bLimAllBwd;
END_IF
```
But this gives me the same "master ignoring slaves limit switch issue", so I do think the `AND Slave.bLimFwd` lines are necessary. If that's true, then doing something like:
```
IF bGantryAlreadyCoupled THEN
	Master.bLimAllFwd := Master.bLimAllFwd AND Slave.bLimAllFwd AND Slave.bLimFwd;
	Master.bLimAllBwd := Master.bLimAllBwd AND Slave.bLimAllBwd AND Slave.bLimBwd;
	
	Slave.bLimAllFwd := Master.bLimAllFwd;
	Slave.bLimAllBwd := Master.bLimAllBwd;
END_IF
```
is somewhat confusing because the `bLimAllFwd` variables are themselves set by the normal limits.

Maybe I'm not seeing the right approach, but given all of this it seems simpler to have this code snippet only set `LimGntFwd` variables and let `FB_MotionStage` set `bLimAllFwd` accordingly.

Reply:
I meant to use the `LimAll` variables on the right hand side instead of expressions like `Slave.bLimGntFwd AND Slave.bLimFwd`

Reply:
So my latest commit caaba7e implements this by setting `bAllForwardEnable` (and Backward/AllEnable) in its own function block. This allows these variables to be set in both `FB_MotionStage` and `FB_GantryAutoCoupling` which was necessary to avoid logic loops where `bAllForwardEnable` was stuck at `FALSE`. @slacAWallace originally suggested using a function as opposed to a function block, but this gets tricky as functions have only one output, meaning I would really need three functions, one of which would set `bAllForwardEnable`, another `bAllBackwardEnable`, and the third `bAllEnable`. I did read that functions can have multiple outputs but its unclear if these can be attributes of a structure.

Really my question though is why should this not be a function block as I'm not immediately seeing the downside of this.